key_handling.c

```diff
@@ -156,17 +178,77 @@ void register_keys(void) {
     }
 
     // send the data
-    err = ubirch_send(CONFIG_UBIRCH_BACKEND_KEY_SERVER_URL, UUID, sbuf->data, sbuf->size, NULL);
-    if(err != ESP_OK) {
-        ESP_LOGE(TAG, "unable to send registration");
+    // TODO: verify response
+    int http_status;
+    if (ubirch_send(CONFIG_UBIRCH_BACKEND_KEY_SERVER_URL, UUID, sbuf->data, sbuf->size, &http_status, NULL, NULL)
+            == UBIRCH_SEND_OK) {
+        if (http_status == 200) {
+            ESP_LOGI(TAG, "successfull sent registration");
+        } else {
+            ESP_LOGE(TAG, "unable to send registration");
+        }
+    } else {
+        ESP_LOGE(TAG, "error while sending registration");
     }
     msgpack_sbuffer_free(sbuf);
-    ESP_LOGI(TAG, "successfull sent registration");
 }
 
 
 void check_key_status(void) {
     if (load_keys() != ESP_OK) {
         create_keys();
     }
-}
\ No newline at end of file
+    // only load default backend key if there is nothing in flash
+    if (load_backend_key() != ESP_OK) {
+        if (set_backend_default_public_key() != ESP_OK) {
+            ESP_LOGE(TAG, "error setting backend pub key");
+        }
+    }
+}
+
+
+esp_err_t set_backend_default_public_key(void) {
+    return set_backend_public_key(CONFIG_UBIRCH_BACKEND_PUBLIC_KEY);
+}
+
+
+esp_err_t set_backend_public_key(const char* keybase64string) {
+    size_t len = strlen(keybase64string);
+    if (len != PUBLICKEY_BASE64_STRING_LENGTH) {
+        ESP_LOGE(TAG, "unexpected base64 string length");
+        return ESP_FAIL;
+    }
+    size_t outputlen = 0;
+    if (mbedtls_base64_decode(server_pub_key, crypto_sign_PUBLICKEYBYTES, &outputlen,
+            (const unsigned char*)keybase64string, len) != 0) {
+        ESP_LOGE(TAG, "decoding base64 failed");
+        return ESP_FAIL;
+    }
+    if (outputlen != crypto_sign_PUBLICKEYBYTES) {
+        ESP_LOGE(TAG, "decoding base64 returned unexpected key length");
+        return ESP_FAIL;
+    }
+    ESP_LOGI(TAG, "setting backend public key");
+    ESP_LOG_BUFFER_HEXDUMP("key", server_pub_key, len, ESP_LOG_INFO);
+    //store the public key
+    esp_err_t err = kv_store("key_storage", "server_key", server_pub_key, crypto_sign_PUBLICKEYBYTES);
+    if (memory_error_check(err)) return err;
+
+    return err;
+}
+
+
+esp_err_t get_backend_public_key(char* buffer, const size_t buffer_size) {
+    unsigned int outputlen = 0;
+    switch (mbedtls_base64_encode((unsigned char*)buffer, buffer_size, &outputlen, server_pub_key, crypto_sign_PUBLICKEYBYTES)) {
```

Comment:
Before accessing `server_pub_key` we should make sure it is in sync with the key in flash memory.